### PR TITLE
ast, topdown: Index comprehensions to avoid unnecessary work

### LINKED ISSUE #2276: Improve performance in queries that perform "group by" operations
Rego doesn't support mutation and there's no explicit "group by" operation. As a result, when policies need to aggregate or group data by some key, they need to use a comprehension and close over values from inside the query. This leads to cross-products in Rego when implementations in other languages would have linear runtime. This can be seen going back all the way to the old scheduler policy. For example:

```rego
# this is from test/scheduler/scheduler_test.go
pods_on_node[node_id] = pds {
    node_name := nodes[node_id].metadata.name
    pds := [p | pods[i].spec.nodeName == node_name; p := pods[i]]
}
```

This rule has O(nodes*pods) runtime. Another common case is inverting an object (which ends up being O(object^2). For example:

```rego
v := obj[_]
ks := [k | some k; v==obj[k]]
```

The latter case is trivial and could be addressed by a built-in function. The scheduler case above is not as easy. E.g., the built-in would have to accept a parameter that indicates how to key any index it builds.

A better solution would be for OPA to identify comprehensions that could and should be indexed (as well as the keys to use for those indices.) During evaluation, the evaluator can lazily build the comprehension index that would allow the query to efficiently lookup the comprehension value based on local variable bindings.

For example, in the scheduler case above, the index would be keyed by "node_name". In the second example, the index would be keyed by "v". The index could be hierarchical to avoid needing to prune lookup results.

Only comprehensions that are safe when excluding the parent query can be indexed. For example:

```rego
x := p[_]
_ = {1 | y = x[_]}  # cannot be indexed because 'x' is not safe
```

Note, in some cases it'll be less efficient to index comprehensions. For example:

```rego
p[x]
ys := {y | q[x][y] = 1}
```

In this case if O(p)>>O(q) then comprehension indexing is effective. OTOH, if O(q)>>O(p) then indexing may be more expensive because all values of 'x' are not known at indexing time. A simple heuristic could be to avoid indexing when keys are used as ref operands because they are probably used to perform constant-time lookups that shrink the search space.

Misc. notes

* Comprehension index must be invalidated by with statements

### PR DESCRIPTION
**Pull request by tsandall**
This commit adds a new kind of indexing to the compiler and topdown to
help avoid recomputing comprehensions. This helps with queries that
perform "group by" operations.

This optimization allows policies to perform group-by/aggregation in
O(n) instead of O(n^2). The optimization works by computing a set of
index keys for the comprehension at compile-time and then computing
the collection once at evaluation-time and indexing the result based
on the keys.

The index keys are variables in the outer query that limit the values
produced by the comprehension. In the simple group-by case these are
the object values themselves. During evaluation, topdown checks if
indexing is possible and builds the index by computing the
comprehension without creating a closure over the outer query. This
computes ALL values in the collection defined by the
comprehension. The results are keyed by the assignments to the
variables indicated in the comprehension index. This way the
comprehension does not have to be recomputed for each set of
assignments in the outer query.

The index is exposed on both the compiler and the query compiler so
that ad-hoc queries can benefit from the indexing as well. This is
important for things like the playground where users may select a rule
body and run it. If that exhibited n^2 behaviour it would be quite
confusing.

In order to be indexed, the comprehension must meet a few
conditions. Importantly, the indexing should not worsen overall
performance. To ensure this, comprehensions containing refs or walk()
calls that include output vars that close over the outer query are not
indexed. This means that if the caller were pushing down assignments
to those vars, OPA will not compute the entire collection.

In the future we can improve the index to cover more kinds of
comprehensions. One improvement that would be particularly nice would
be to allow the comprehension index to close over specific local
variables in the parent scope. This would let us build the index in
more cases--however, the analysis would need to be careful to take
into account the count of closure variables. Variables with multiple
assignments would be poor candidates.

Benchmark results (before, O(n^2) runtime):

BenchmarkComprehensionIndexing/10-16 	   13831	     85821 ns/op
BenchmarkComprehensionIndexing/100-16         	     208	   5662625 ns/op
BenchmarkComprehensionIndexing/1000-16        	       2	 549295038 ns/op

Benchmark results (after, O(n) runtime):

BenchmarkComprehensionIndexing/10-16 	   35809	     33369 ns/op
BenchmarkComprehensionIndexing/100-16         	    3756	    274546 ns/op
BenchmarkComprehensionIndexing/1000-16        	     438	   2725152 ns/op

Fixes #2276

Signed-off-by: Torin Sandall <torinsandall@gmail.com>

### CODE REVIEW

File: docs/content/policy-performance.md [hunk 1]
```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
...
+  ports := [port |
+    some j
+    input.exposed[j].interface == intf
+    port := input.exposed[j].port
+  ]
+}
+```
+
+Comprehension indices can be built for any comprehension (nested or not) if the following restrictions are met:
+
```
Comment (timothyhinrichs) [inline docs/content/policy-performance.md:219]: Would be helpful if there were a higher-level description of at least the common case.  Maybe something like "an iteration with a comprehension that is used to find a collection of objects that are equal to (or more generally compared to) the iteration variable".    Not happy with the wording, but something similar.

Perhaps also an example of what does NOT get indexed but might seem like it should.  Edit: you did this below.  Nice!

File: docs/content/policy-performance.md [hunk 2]
```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
...
+The following examples show cases that are NOT indexed:
+
+```rego
+not_indexed_because_missing_assignment {
+  x := input[_]
+  [y | some y; x == input[y]]
+}
+
+not_indexed_because_because_includes_with {
```
Comment (timothyhinrichs) [inline docs/content/policy-performance.md:234]: because_because

File: docs/content/policy-performance.md [hunk 3]
```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
...
+  x := input[_]
+  ys := [y | some y; x := input[y]] with input as {}
+}
+
+not_indexed_because_because_negated {
```
Comment (timothyhinrichs) [inline docs/content/policy-performance.md:239]: because_because

File: docs/content/policy-performance.md [hunk 4]
```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
...
+  x := input[_]
+  not data.arr = [y | some y; x := input[y]]
+}
+
+not_indexed_because_because_safety {
```
Comment (timothyhinrichs) [inline docs/content/policy-performance.md:244]: because_because

File: docs/content/policy-performance.md [hunk 5]
```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
...
+  obj := input.foo.bar
+  x := obj[_]
+  ys := [y | some y; x == obj[y]]
+}
+
+not_indexed_because_because_reference_output[y] {
```
Comment (timothyhinrichs) [inline docs/content/policy-performance.md:250]: because_because

File: topdown/eval.go [hunk 6]
```diff
@@ -859,7 +869,107 @@ func (e *eval) biunifyComprehension(a, b *ast.Term, b1, b2 *bindings, swap bool,
 		return e.biunifyComprehensionObject(a, b, b1, b2, iter)
 	}
 
-	return fmt.Errorf("illegal comprehension %T", a)
+	return internalErr(e.query[e.index].Location, "illegal comprehension type")
+}
+
+func (e *eval) indexComprehension(a *ast.Term) (*ast.Term, error) {
+
+	index := e.comprehensionIndex(a)
```
Comment (patrick-east) [inline topdown/eval.go:877]: nit: Having both `indexComprehension` and `comprehensionIndex` on `eval` seems less than ideal, could we distinguish the names more?

File: docs/content/policy-performance.md (hunk 1 above)
Comment (tsandall) [reply docs/content/policy-performance.md:219]: Changed the description to this:

```

Indices can be built for comprehensions (nested or not) that generate collections (i.e., arrays, sets, or objects)
based on variables in an outer query. In the example above:

* `intf` is the variable in the outer query.
* `[port | some j; input.exposed[j].interface == intf; port := input.exposed[j].port]` is the comprehension.
* `ports` is the variable the collection is assigned to.
```

Then list out the conditions/exceptions.

File: ast/compile.go [hunk 7]
```diff
@@ -1462,6 +1506,189 @@ func (qc *queryCompiler) rewriteWithModifiers(qctx *QueryContext, body Body) (Bo
 	return body, nil
 }
 
+func (qc *queryCompiler) buildComprehensionIndices(qctx *QueryContext, body Body) (Body, error) {
+	// NOTE(tsandall): The query compiler does not have a metrics object so we
+	// cannot record index metrics currently.
```
[the remark this answers is not in the document]
Comment (tsandall) [reply ast/compile.go:1511]: We can although I'm also fine just waiting until this shows up again. Haven't really seen a need for it to date.

File: topdown/eval.go (hunk 6 above)
Comment (tsandall) [reply topdown/eval.go]: Renamed to `indexComprehension` to `buildComprehensionCache`.

File: ast/compile.go [hunk 8]
```diff
@@ -1462,6 +1506,189 @@ func (qc *queryCompiler) rewriteWithModifiers(qctx *QueryContext, body Body) (Bo
...
+	_ = buildComprehensionIndices(qc.compiler.GetArity, ReservedVars, body, qc.comprehensionIndices)
+	return body, nil
+}
+
+// ComprehensionIndex specifies how the comprehension term can be indexed. The keys
+// tell the evaluator what variables to to use for indexing. In the future, the index
```
Comment (ashutosh-narkar) [inline ast/compile.go:1517]: Nit: `to to`

File: ast/compile.go [hunk 9]
```diff
@@ -1462,6 +1506,189 @@ func (qc *queryCompiler) rewriteWithModifiers(qctx *QueryContext, body Body) (Bo
...
+	//	[true | data.y[z] = x]     # safe to evaluate w/o outer body
+	//	[true | data.y[z] = x[0]]  # NOT safe to evaluate because 'x' would be unsafe.
+	//
+	// By identifying output variables in the body we also know what to index on by
+	// intersecting with candidate variables from the outer query.
+	//
+	// For example:
+	//
+	//	x = data.foo[_]
+	//	[y | data.bar[y] = x]      # index on 'x'
```
Comment (ashutosh-narkar) [inline ast/compile.go:1587]: Should this example `[y | data.bar[y] = x]` be of the form `<var> = <comprehension>` ? 

Also why do we skip single term comprehensions eg `[v | input[i] = v]` ? Can't they be rewritten as equality expressions ?

File: ast/compile_test.go [hunk 10]
```diff
@@ -3377,6 +3377,224 @@ func TestCompilerWithStageAfterWithMetrics(t *testing.T) {
...
+				}`,
+		},
+		{
+			note: "skip: due to negation",
+			module: `
+				package test
+
+				p {
+					v = input[i]
+					not [j | input[j] = v] # skip because of with modifier
```
Comment (ashutosh-narkar) [inline ast/compile_test.go:3458]: Nit: `# skip because of negation`

File: ast/compile.go (hunk 8 above)
Comment (tsandall) [reply ast/compile.go]: Fixed.

File: ast/compile.go (hunk 9 above)
Comment (tsandall) [reply ast/compile.go]: Fixed. The reason it skips single term exprs is that the collection generated by the comprehension is not assigned anywhere (so there's no point in indexing it.)

File: ast/compile_test.go (hunk 10 above)
Comment (tsandall) [reply ast/compile_test.go]: Fixed.

**Pull request merged**